### PR TITLE
Improvements to the display of the experience, todo and task charts

### PR DESCRIPTION
**Pull request by ltorroba**
1) Fixes issue #6679, which referred to the EXP chart getting cropped.
2) Added some bottom padding to the EXP chart so that the tags are
further apart from the chart itself (requires rebuilding stylus)
3) Added event handlers for the dynamic resizing of the charts. The
event handlers use the same data every time and are only created when a
chart is first toggled, as to reduce overhead.
4) Moved the location of the tasks chart into the .task-text, so that
the habit +/- buttons are now fully triggerable.

I’ll add more details on the implementation below soon.

### CODE REVIEW

**Comment by ltorroba** [comment]
I picked up issue #6679, and noticed that alongside fixing the clipping problem this would be a good chance to fix a few other bugs related to charts. They were listed above. On the whole, I believe this PR would make the chart somewhat prettier, especially when it comes to dynamic/flexible container sizes.

## Charts currently at desktop resolution (Safari @ Mac):

### EXP



### Todos



### Task

Portion of the `+` sign becomes unclickable as the chart floats above it, See next section for the specific region,

_The same behaviour was observed on Chrome on the original issue, and confirmed by tests._

## Charts currently, at reduced resolution (Chrome @ Mac):

### EXP



### Todos



### Task

_Moreover, bounded region becomes unclickable_

_I also verified this on Firefox and Safari._

## Charts after these commits, at desktop resolution (Chrome @ Mac)

### EXP



### Todos



### Task


_Also tested on Firefox and Safari._

## Charts after these commits, at reduced resolution (Chrome @ Mac)

### EXP



### Todos



### Task



# The only problem

Only exception is at very specific screen sizes (e.g.: task column div width is between 271 and 273 pixels) on Safari and Firefox, where the following may occur:


I have tried different workarounds but it seems to be a library and browser issue, so we are not talking anything neat. Any ideas? Should we fix this, and/or try to report it as an issue with the Google Charts API itself?

**Comment by keysmashes** [comment]
Is it even worth showing the legend? It seems like it should be self-evident, and 'score' isn't very informative anyway.

File: website/public/js/controllers/rootCtrl.js [hunk 1]
```diff
@@ -169,7 +169,13 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
 
     $rootScope.charts = {};
     $rootScope.toggleChart = function(id, task) {
-      var history = [], matrix, data, chart, options;
+      if($rootScope.charts[id] === undefined) {
+        $(window).resize(function() {
```
Comment (crookedneighbor) [inline website/public/js/controllers/rootCtrl.js]: Window resize functions should be debounced.

File: website/public/js/controllers/rootCtrl.js [hunk 2]
```diff
@@ -169,7 +169,13 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
 
     $rootScope.charts = {};
     $rootScope.toggleChart = function(id, task) {
-      var history = [], matrix, data, chart, options;
+      if($rootScope.charts[id] === undefined) {
```
Comment (crookedneighbor) [inline website/public/js/controllers/rootCtrl.js]: A simple `if (!$rootScope.charts[id])` should suffice here.

File: website/views/options/social/tavern.jade [hunk 3]
```diff
@@ -83,7 +83,7 @@
                 a.label.label-contributor-1(ng-click='toggleUserTier($event)')=env.t('tier') + ' 1 (' + env.t('friend') + ')'
                 div
                   p
-                    span.achievement.achievement-firefox
+                    span.achievement.achievement-boot
```
Comment (crookedneighbor) [inline website/views/options/social/tavern.jade:86]: This doesn't look related to this PR.

File: website/public/js/controllers/rootCtrl.js [hunk 4]
```diff
@@ -184,23 +190,62 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
...
       matrix = [[env.t('date'), env.t('score')]];
       _.each(history, function(obj) {
         matrix.push([moment(obj.date).format(User.user.preferences.dateFormat.toUpperCase().replace('YYYY','YY') ), obj.value]);
       });
       data = google.visualization.arrayToDataTable(matrix);
+
+      $rootScope.drawChart(id, data);
+    };
+
+    $rootScope.drawChart = function(id, data, options) {
```
Comment (crookedneighbor) [inline website/public/js/controllers/rootCtrl.js]: Does this need to be exposed onto `$rootScope`? Looks like it is only called internally.

**Comment by ltorroba** [comment]
@anowlcalledjosh @crookedneighbor Sorry for the delay, but here are the commits associated with your suggestions. Let me know if there's anything else to be done. 😄

**Comment by Alys** [comment]
@zesme Thanks for that! Just so you know, we won't be able to review this until late this month because we have a code freeze now while we finish the final stages of version 3 of Habitica's API. If you'd like more information, see the new "API Changes" section in Guidance for Blacksmiths at http://habitica.wikia.com/wiki/Guidance_for_Blacksmiths#API_Changes and you're welcome to ask questions in the Aspiring Coders guild.

**Comment by ltorroba** [comment]
@Alys No worries, I had heard of the code freeze but, ironically, this To-Do had gotten quite red in my list, so I felt it was time to get it over with. As soon as the new API gets released I will get working on the merging.

File: website/views/shared/tasks/task.jade [hunk 5]
```diff
@@ -14,4 +14,4 @@ li(id='task-{{::task.id}}',
 
     include ./edit/index
 
-  div(class='{{obj._id}}{{task.id}}-chart', ng-show='charts[obj._id+task.id]')
+
```
Comment (awitherow) [inline website/views/shared/tasks/task.jade]: is this necessary?

File: website/public/js/controllers/rootCtrl.js [hunk 6]
```diff
@@ -184,23 +184,73 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
           $rootScope.charts[id] = (history.length == 0) ? false : !$rootScope.charts[id];
           if (task && task._editing) task._editing = false;
       }
+
+      if($rootScope.charts[id]) {
```
Comment (awitherow) [inline website/public/js/controllers/rootCtrl.js]: I see a lot of cases with spaces between the if and the parenthesis, and also a lot of cases not with. Is there a code styleguide that is being followed?

Comment (crookedneighbor) [inline website/public/js/controllers/rootCtrl.js]: We've instituted eslint for the server code. We're about to do a rewrite of the client code, so we haven't bothered including it in the linting, since it will be replaced soon.

File: website/views/shared/tasks/task.jade (hunk 5 above)
Comment (ltorroba) [inline website/views/shared/tasks/task.jade]: Good point! Not quite sure what was going on in my head when I removed that. I just made a push to bring it back.

**Comment by ltorroba** [comment]
Fixed conflicts with base branch

File: website/client/js/controllers/rootCtrl.js [hunk 7]
```diff
@@ -185,23 +185,73 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
           $rootScope.charts[id] = (history.length == 0) ? false : !$rootScope.charts[id];
           if (task && task._editing) task._editing = false;
       }
+
+      if($rootScope.charts[id]) {
+        var handleResize = _.debounce(function() {
+          drawChart(id, data);
+        }, 300);
+
+        $(window).resize(handleResize);
```
Comment (crookedneighbor) [inline website/client/js/controllers/rootCtrl.js]: This is adding a new event listener to the window resize every time the chart is toggled. Meaning if you toggle back and forth a few times, you add a new listener each time and the old listeners do not get cleaned up. This will be a performance issue.

File: website/client/js/controllers/rootCtrl.js [hunk 8]
```diff
@@ -185,23 +185,73 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
...
       });
       data = google.visualization.arrayToDataTable(matrix);
+
+      drawChart(id, data);
+    };
+
+    function drawChart(id, data, options) {
+      var chart, width;
+
+      switch(id) {
```
Comment (crookedneighbor) [inline website/client/js/controllers/rootCtrl.js]: Let's use `if/else if/else` here instead of a switch.

**Comment by crookedneighbor** [comment]
Sorry for the super long delay on reviewing this. It looks 95% of the way there. The one big problem with it is that it's setting window resize listeners without cleaning up after itself.

**Comment by ltorroba** [comment]
@crookedneighbor No worries! I am kinda busy this week, but as soon as I get a chance I will get everything sorted.

**Comment by crookedneighbor** [comment]
@zesme Status report on this?

**Comment by ltorroba** [comment]
@crookedneighbor I'll be taking a look at this today or within the next two days by the latest. Sorry about the delay!

**Comment by ltorroba** [comment]
@crookedneighbor Made the changes you asked for. 66922a6 seems kinda hacky to me, but I couldn't figure out a better way to implement it. If you (or anyone else) has any ideas I'll happily improve it!

**Comment by crookedneighbor** [comment]
Did a tiny adjustment: e3b9636c42d2641fe038df6226ad548e88d4f1b3

Basically, just set one listener on the window and looped through the resize functions for each active chart.

Notated toward your next tier.
